javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -44,50 +71,134 @@ class ArrayList[E] private (private[ArrayList] val inner: js.Array[E])
+    if (targetPureWasm) {
+      _size
+    } else {
+      innerJS.length
+    }
 
   override def clone(): AnyRef =
-    new ArrayList(inner.jsSlice(0))
+    if (targetPureWasm) {
+      new ArrayList(innerWasm, size())
+    } else {
+      new ArrayList(innerJS.jsSlice(0), size())
+    }
 
   def get(index: Int): E = {
     checkIndexInBounds(index)
-    inner(index)
+    if (targetPureWasm) {
+      innerWasm(index).asInstanceOf[E]
+    } else {
+      innerJS(index)
+    }
   }
 
   override def set(index: Int, element: E): E = {
     val e = get(index)
-    inner(index) = element
+    if (targetPureWasm) {
+      innerWasm(index) = element
+    } else {
+      innerJS(index) = element
+    }
     e
   }
 
   override def add(e: E): Boolean = {
-    inner.push(e)
+    if (targetPureWasm) {
+      add(_size, e)
+    } else {
+      innerJS.push(e)
+    }
     true
   }
 
   override def add(index: Int, element: E): Unit = {
     checkIndexOnBounds(index)
-    inner.splice(index, 0, element)
+    if (targetPureWasm) {
+      if (size() >= innerWasm.length) expand()
+      var i = size()
+      while (i > index) {
+        innerWasm(i) = innerWasm(i - 1)
+        i -= 1
+      }
+      innerWasm(i) = element
+      _size += 1
+    } else {
+      innerJS.splice(index, 0, element)
+    }
   }
 
   override def remove(index: Int): E = {
     checkIndexInBounds(index)
-    arrayRemoveAndGet(inner, index)
+    if (targetPureWasm) {
+      val removed = innerWasm(index).asInstanceOf[E]
+      for (i <- index until (size() - 1)) {
+        innerWasm(i) = innerWasm(i + 1)
+      }
+      innerWasm(size - 1) = null
+      _size -= 1
+      removed
+    } else {
+      arrayRemoveAndGet(innerJS, index)
+    }
   }
 
   override def clear(): Unit =
-    inner.length = 0
+    if (targetPureWasm) {
+      for (i <- (0 until size())) {
+        innerWasm(i) = null
+      }
+      _size = 0
+    } else {
+      innerJS.length = 0
+    }
 
   override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
-    c match {
-      case other: ArrayList[_] =>
-        inner.splice(index, 0, other.inner.toSeq: _*)
-        other.size() > 0
-      case _ => super.addAll(index, c)
+    if (targetPureWasm) {
+      checkIndexOnBounds(index)
+      var i = index
+      val iter = c.iterator()
+      while (iter.hasNext()) {
+        add(i, iter.next())
+        i += 1
+      }
+      !c.isEmpty()
+    } else {
+      c match {
+        case other: ArrayList[_] =>
+          innerJS.splice(index, 0, other.innerJS.toSeq: _*)
+          other.size() > 0
+        case _ => super.addAll(index, c)
+      }
     }
   }
 
   override protected def removeRange(fromIndex: Int, toIndex: Int): Unit =
-    inner.splice(fromIndex, toIndex - fromIndex)
+    if (targetPureWasm) {
+      if (fromIndex != toIndex) {
+        if (fromIndex < 0 || fromIndex >= size() ||
+            toIndex > size() || toIndex < fromIndex) {
+          throw new IndexOutOfBoundsException()
```

Comment:
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ArrayList.html#removeRange(int,int)

Maybe JS implementation should also throw an exception here?

Reply:
Ah yes. The bounds checks should be common to the two implementations. (also in `addAll`, actually?)

Looks like we're missing tests :p